Enable forwarding of notifications

pvollan opened the pull request:
#### 1f0b1256ce864aee7b8d7142c8177f1671c8c83f
<pre>
Enable forwarding of notifications
<a href="https://bugs.webkit.org/show_bug.cgi?id=270003">https://bugs.webkit.org/show_bug.cgi?id=270003</a>
<a href="https://rdar.apple.com/123524381">rdar://123524381</a>

Reviewed by Brent Fulgham.

Enable forwarding of notifications from the UI process to the WebContent process.

* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerNotificationObservers):
(WebKit::WebProcessPool::unregisterNotificationObservers):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::postNotification):
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:

Canonical link: <a href="https://commits.webkit.org/275342@main">https://commits.webkit.org/275342@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/191a0627903efee5ae7515581bf337f33543270f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41549 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20563 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43927 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44118 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37641 "Built successfully") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/23724 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17893 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34359 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42123 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17491 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35785 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15019 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15207 "layout-tests (failure)") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45489 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37737 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37111 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40878 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16364 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13443 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39286 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17983 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18039 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5559 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17627 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/338b447cbb689ff748136d5d3c14cc0f5c3f7433. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25029

brentfulgham commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -713,11 +721,13 @@ static void logProcessPoolState(const WebProcessPool& pool)
     m_notifyTokens = WTF::compactMap(notificationMessages, [this](const ASCIILiteral& message) -> std::optional<int> {
         int notifyToken = 0;
-        auto status = notify_register_dispatch(message, &notifyToken, dispatch_get_main_queue(), ^(int token) {
+        auto status = notify_register_dispatch(message, &notifyToken, dispatch_get_main_queue(), [this, message](int token) {
```
Do we need to Retain `this` to avoid it being collected when the notification runs?

pvollan replied:
That is a good point! I updated the patch to use a weak pointer, instead.

Thanks for reviewing!

brentfulgham reviewed (approved):
Looks reasonable. r=me

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/80ffe2b8ab734093e6671bea5f1ebb6a5cebe1f5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25029

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/50db2d3647dee6ebc745f623e177109b6bca32b0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25029

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/83c5bc412ccdc16ef79e1b40af75c44239ffd101. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25029

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -703,21 +706,30 @@ static void logProcessPoolState(const WebProcessPool& pool)
-        auto status = notify_register_dispatch(message, &notifyToken, dispatch_get_main_queue(), ^(int token) {
-            if (!m_processes.isEmpty()) {
+        auto status = notify_register_dispatch(message, &notifyToken, dispatch_get_main_queue(), [weakThis, message](int token) {
+            RefPtr strongThis = weakThis.get();
```
We don't use the `strong` naming with C++ classes. We only use `strong` in OjbC. Please rename to `protectedThis`.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -728,11 +740,14 @@ static void logProcessPoolState(const WebProcessPool& pool)
-        RetainPtr<NSObject> observer = [[NSNotificationCenter defaultCenter] addObserverForName:message object:nil queue:[NSOperationQueue currentQueue] usingBlock:^(NSNotification *notification) {
-            if (!m_processes.isEmpty()) {
+    m_notificationObservers = WTF::compactMap(nsNotificationMessages, [weakThis = WeakPtr { *this }](NSString* message) -> RetainPtr<NSObject>  {
+        RetainPtr<NSObject> observer = [[NSNotificationCenter defaultCenter] addObserverForName:message object:nil queue:[NSOperationQueue currentQueue] usingBlock:[weakThis, message](NSNotification *notification) {
```
Do we really need `<NSObject>` here?

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -728,11 +740,14 @@ static void logProcessPoolState(const WebProcessPool& pool)
-            if (!m_processes.isEmpty()) {
+    m_notificationObservers = WTF::compactMap(nsNotificationMessages, [weakThis = WeakPtr { *this }](NSString* message) -> RetainPtr<NSObject>  {
+        RetainPtr<NSObject> observer = [[NSNotificationCenter defaultCenter] addObserverForName:message object:nil queue:[NSOperationQueue currentQueue] usingBlock:[weakThis, message](NSNotification *notification) {
+            RefPtr strongThis = weakThis.get();
```
`protectedThis`

pvollan replied:
Ah, good points, fixed in latest patch.

Thanks for reviewing!

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/191a0627903efee5ae7515581bf337f33543270f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/25029

pvollan committed:
Enable forwarding of notifications
https://bugs.webkit.org/show_bug.cgi?id=270003
rdar://123524381

Reviewed by Brent Fulgham.

Enable forwarding of notifications from the UI process to the WebContent process.

* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerNotificationObservers):
(WebKit::WebProcessPool::unregisterNotificationObservers):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::postNotification):
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:

Canonical link: https://commits.webkit.org/275342@main

webkit-commit-queue commented:
Committed 275342@main (1f0b1256ce86): <https://commits.webkit.org/275342@main>

Reviewed commits have been landed. Closing PR #25029 and removing active labels.
